### PR TITLE
Add CollisionShape3D custom debug colors

### LINKED ISSUE #906: More configurable/custom colors for visible collision shapes debug option (already implemented in 2D)
**Describe the project you are working on:**
N/A (generic issue)

**Describe the problem or limitation you are having in your project:**
Difficulty to pick out collision shapes, especially when they overlap (it's more pronounced in 2D than 3D)
Tangent: collision shapes and navmeshes/navpolys coexisting in the same project exacerbate the problem even further.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
Different collision shape colors would help differentiate between shapes, esp. overlapping ones.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
In the original issue, @Calinou suggested either randomly selecting colors or exposing a custom_color option.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
Not possible to work around AFAIK.

**Is there a reason why this should be core and not an add-on in the asset library?:**
Can assets even touch debug collision shape setting in any meaningful way? IIRC no.

Original issue: https://github.com/godotengine/godot/issues/31283

### PR DESCRIPTION
**Pull request by BattyBovine**
This allows changing the display colour of a collision shape preview on a per-shape basis. It also adds the ability to display a solid coloured preview of a collision shape.

Closes https://github.com/godotengine/godot-proposals/issues/906

This is a feature that I had been working on independently before realising that https://github.com/godotengine/godot/pull/78273 already exists. My implementation is currently more complete, in that all collision shapes are supported, including height maps and convex/concave collision shapes. I was hesitant to make a duplicate pull request for this feature, but I was told I could do so in the comments of the other pull request.

Screenshot - 4_8_2024 , 7_44_08 PM

Screenshot - 4_13_2024 , 4_42_33 PM

### CODE REVIEW

**Comment by Zireael07** [comment]
Honestly I NEED one of those to finally make it in!

File: doc/classes/CollisionShape3D.xml
Comment (Mickeon) [inline doc/classes/CollisionShape3D.xml:34]: The note is misleading as if to hide how this actually works. What it entails is probably something more like this:
```suggestion
			The collision shape's debug color. If this value is set to [code]Color(0, 0, 0, 0)[/code], the value of [member EditorSettings.editors/3d_gizmos/gizmo_colors/shape] is used, instead.
```

File: scene/resources/3d/capsule_shape_3d.cpp
Comment (Mickeon) [inline scene/resources/3d/capsule_shape_3d.cpp:157]: As per comments standard.
```suggestion
	// Bottom hemisphere.
```

File: scene/resources/3d/capsule_shape_3d.cpp
Comment (Mickeon) [inline scene/resources/3d/capsule_shape_3d.cpp:120]:
```suggestion
	// Cylinder.
```

File: scene/resources/3d/capsule_shape_3d.cpp
Comment (Mickeon) [inline scene/resources/3d/capsule_shape_3d.cpp:83]:
```suggestion
	// Top hemisphere.
```

File: scene/resources/3d/box_shape_3d.cpp
Comment (Mickeon) [inline scene/resources/3d/box_shape_3d.cpp:153]:
```suggestion
	// Top + bottom.
```

File: scene/resources/3d/box_shape_3d.cpp
Comment (Mickeon) [inline scene/resources/3d/box_shape_3d.cpp:160]:
```suggestion
			// Top.
```

File: scene/resources/3d/box_shape_3d.cpp
Comment (Mickeon) [inline scene/resources/3d/box_shape_3d.cpp:165]:
```suggestion
			// Bottom.
```

File: scene/resources/3d/box_shape_3d.cpp
Comment (Mickeon) [inline scene/resources/3d/box_shape_3d.cpp:108]: There's more like this in the same file, I'm not sure it's productive to list all of them
```suggestion
	// Left + right.
```

File: doc/classes/CollisionShape3D.xml
Comment (BattyBovine) [reply doc/classes/CollisionShape3D.xml]: It was originally written that way to keep it consistent with CollisionShape2D. Your suggestion reads a lot more like a proper description though, so I've rewritten it similarly.

File: scene/resources/3d/capsule_shape_3d.cpp
Comment (BattyBovine) [reply scene/resources/3d/capsule_shape_3d.cpp]: Those comments were from the parts of Godot's code that the mesh generation code came from. I copied and modified that code because I was concerned about the extra data created by those functions making mesh generation slower than it would need to be. In retrospect, this is probably not a concern, so all of these functions were rewritten. They now use the `create_mesh_array` functions for all shape primitives that have it, with added vertex colour data.

**Review by Calinou — CHANGES_REQUESTED** [review]
Tested locally (rebased on top of `master` 292e50e17e3f6e2509d3178a00204f964a907460), the base feature mostly works so far. This is encouraging :slightly_smiling_face: 

I noticed some issues though:

**Testing project:** [test_pr_90644.zip](https://github.com/user-attachments/files/15791900/test_pr_90644.zip)

- At runtime, instead of following the properties set for each shape in the editor, the collision shape color is defined by the *last* collision shape that is parented to a PhysicsBody3D in the scene tree:

Editor | Runtime
-|-
Screenshot_20240611_180922 | Screenshot_20240611_180914

- The default shape opacity is lower, which makes their outlines less visible than in `master`;

`master` | This PR
-|-
Screenshot_20240611_181128 | Screenshot_20240611_181219

I suggest leaving the default color's opacity at 255 to avoid this. That said, even with opacity set to 255, collision shape lines become slightly less opaque than in `master`:

Screenshot_20240611_181328

- Trimesh collision shape drawing seems to break under certain conditions, as seen in the above screenshot.

**Comment by BattyBovine** [comment]
>     * The default shape opacity is lower, which makes their outlines less visible than in `master`;
This is an issue I noticed as well, and I did my best to compensate for it. I'll see about fixing it further, but as far as I can tell, this is an issue brought about by using vertex colours as albedo. No matter what I tried, the opacity was always significantly lower, no matter whether `FLAG_SRGB_VERTEX_COLOR` was set, or how I compensated for it myself, or whether opacity was set to the maximum. This might not be a problem I can fully solve with this PR.

I will look into this though, and the other issues are probably easily resolved, too.

**Comment by BattyBovine** [comment]
Sorry for the immediate double-post, but I think I figured all of these issues out:

> At runtime, instead of following the properties set for each shape in the editor, the collision shape color is defined by the _last_ collision shape that is parented to a PhysicsBody3D in the scene tree:

This is actually not what was happening. It just appeared that way due to a different issue. The actual problem was that the custom colours were being set properly on the `CollisionShape3D`, but the `Shape3D` resource attached to these nodes were not unique, so they were secretly all sharing the same colour regardless of the `CollisionShape3D` property. I couldn't think of a way to make the `CollisionShape3D` nodes all update to reflect this without massively overcomplicating things, so I made the decision to make the debug colour properties part of the `Shape3D` resource instead. I didn't do this initially since I wanted to keep things consistent with the `CollisionShape2D` node, but this seems like a reasonable solution.

> The default shape opacity is lower, which makes their outlines less visible than in `master`;

The issue here seems to just be a mistake I made and never noticed. Debug lines were having their opacity multiplied by 0.25 inside the `Shape3D` resource, for reasons I can't remember in the slightest. I simply removed that, and now everything seems much better.

> Trimesh collision shape drawing seems to break under certain conditions, as seen in the above screenshot.

I have not been able to reproduce this since making the changes described above, so I think whatever was happening there was related to either the lower-than-intended opacity, or is fixed due to the `Shape3D` resource handling its own colour settings.

compare

**Comment by Mickeon** [comment]
This failed the **Linux / Editor w/ Mono** check and I am not sure why?

**Comment by Calinou** [comment]
I recall C# failing on CI a few weeks ago, so a rebase *should* fix the issue.

**Review by Mickeon — APPROVED** [review]
Documentation is completely fine now. Something similar to this feature has been **heavily** requested, too.

The PR and commit title should probably be changed, as this PR modifies **Shape3D**.

https://github.com/godotengine/godot/pull/78273 also exists, but it's not as maintained, at all. I am to assume this PR would essentially make it redundant, but there are benefits in tying the debug color to **CollisionShape3D**, so it's worth keeping in mind.

**Comment by BattyBovine** [comment]
It would make more sense to keep the colour options in `CollisionShape3D`, so I'd like to make that happen again. Moving it to `Shape3D` is an easy and clean fix for unique resources causing the colour settings to desync between nodes, but it's not an ideal fix. I'll take a look at the other PR later to see how it handles colour changes with non-unique `Shape3D` resources on `CollisionShape3D` nodes.

**Comment by BattyBovine** [comment]
In light of the recent release of Godot 4.3, I have started working on this again. I finally figured out a method for updating all `CollisionShape3D` properties properly when one `Shape3D` resource is shared among many nodes. I was hoping that the other pull request for this feature would help, but it seems that had the same issue with not properly handling non-unique resources. As such, I came up with my own solution. It's not the most elegant, and I hope to improve it over time, but it does ensure that changes are synced correctly and it seems to work just fine.

File: doc/classes/CollisionShape3D.xml
Comment (Mickeon) [inline doc/classes/CollisionShape3D.xml:36]:
```suggestion
			If [code]true[/code] displays a solid preview in addition to the wireframe.
```
Or
```suggestion
			If [code]true[/code], when the shape is displayed, it will also show a solid color in addition to its wireframe.
```

File: scene/3d/physics/collision_shape_3d.cpp
Comment (Mickeon) [inline scene/3d/physics/collision_shape_3d.cpp:105]: Somewhat nasty, given this snippet is repeated in `set_debug_color()` itself. You may want to have a separate function that does this specific check and updates the debug color.

Comment (BattyBovine) [reply scene/3d/physics/collision_shape_3d.cpp]: I've reworked a lot of situations like this to reduce the many questionable copy-pastes throughout the class. The code for ensuring the "placeholder" colour value resets to the default collision colour as needed is much simpler now.

**Comment by Yogoda** [comment]
Hello, what is going on with this issue? Is there something we can do to help? I really would like to see this in 4.4

**Comment by Zireael07** [comment]
@Yogoda  Same here!

**Comment by Mickeon** [comment]
Awaiting blocking review from @Calinou unless someone else would like to chime in.

**Review by Calinou — CHANGES_REQUESTED** [review]
Tested locally (rebased on top of `master` https://github.com/godotengine/godot/commit/5ccbf6e4c794a4e47456edd9434b75fcd6096a2f) with [Truck Town](https://github.com/godotengine/godot-demo-projects/tree/master/3d/truck_town), there's an issue where **Visible Collision Shapes** has no effect when running a project now (it works in the same conditions in 4.3.stable). The same happens when using `--debug-collisions` on the command line.

The debug collision shapes are correctly rendered in the editor.

PS: I wonder if **Debug Fill** should be enabled by default, so that 3D behavior matches 2D. Having a visible fill helps a lot when trying to spot intersections between shapes, so it's important to have good usability out of the box.

**Comment by BattyBovine** [comment]
I've looked into this a little bit. It seems to be an issue with inherited scenes that don't change the default collision colours. It might not be updating the default debug colour correctly, and is rendering as the hidden default of black with full transparency. I'll look into this more later.

And as long as everyone agrees that it's a good idea, enabling debug fill by default is fine by me, and is an easy change.

**Review by radiantgurl — APPROVED** [review]
Tested and working on  4.3.1 for weeks with no issues

**Comment by BattyBovine** [comment]
I think I figured out the issue. In some cases, the proper colours were simply never set, causing them to appear invisible. This seemed to happen most commonly with instanced scenes, which is why Truck Town showed no collision. I pushed a change to fix this issue, and I'm going to investigate it more later to see if more changes are needed.

**Review by Calinou — APPROVED** [review]
Tested locally, it works as expected. Code looks good to me.

image

I suggest enabling Debug Fill by default, then this should be good to merge.

File: doc/classes/CollisionShape3D.xml
Comment (Calinou) [inline doc/classes/CollisionShape3D.xml:33]:
```suggestion
			The collision shape color that is displayed in the editor, or in the running project when [b]Debug > Visible Collision Shapes[/b] is checked at the top of the editor. If this is reset to its default value of [code]Color(0, 0, 0, 0)[/code], the value of [member ProjectSettings.debug/shapes/collision/shape_color] will be used instead.
```

**Review by AThousandShips — APPROVED** [review]

**Comment by BattyBovine** [comment]
All suggested changes are resolved, and Debug Fill is now enabled by default.

**Review by fire — APPROVED** [review]
Approving since all the testers approved and the newer changes were only documentation.

**Comment by fire** [comment]
Need to wait for the tests to pass before merge queue adding.

**Comment by akien-mga** [comment]
Rebased to solve merge conflicts.

**Comment by Repiteo** [comment]
Thanks! Congratulations on your first merged contribution! 🎉

**Comment by Mickeon** [comment]
Could you share your own script? It may expose the reason for the rendering to behave as is in the current PR.

**Comment by BattyBovine** [comment]
If the only thing that needs to change here is to use `TRANSPARENCY_ALPHA_DEPTH_PRE_PASS`, then that's a very easy fix. As long as it doesn't cause any other issues somehow, I can make another pull request for that. I'll look into this soon.

**Comment by smix8** [comment]
Other debug has an "x-ray" property for the debug that controls if part of the mesh (e.g. edges only) should have depth cull or not. Might want to go into that direction with settings instead of hard-coding only one material property.

**Comment by Calinou** [comment]
We should implement [`glPolygonOffset()`](https://registry.khronos.org/OpenGL-Refpages/gl4/html/glPolygonOffset.xhtml) emulation in materials so that gizmo materials can make use of it. This will allow them to always be drawn in front of overlapping meshes without Z-fighting, but while still being able to make use of the usual depth sorting. This will also help implement https://github.com/godotengine/godot-proposals/issues/1000.

This likely involves exposing a constant depth offset property in BaseMatereial3D.

Engines like DarkPlaces use this to prevent Z-fighting when drawing mesh-based decals and shape debug drawing:

image

**Comment by Calinou** [comment]
https://github.com/godotengine/godot/pull/100211 implements something akin to the aforementioned `r_showcollisionbrushes_polygonoffset`. Please test :slightly_smiling_face:

**Pull request merged**
